code/game/atoms_movable.dm

```diff
@@ -463,25 +463,41 @@
 		var/mob/living/M = hit_atom
 		M.hitby(src, speed)
 
-	else if(isobj(hit_atom)) // Thrown object hits another object and moves it
-		var/obj/O = hit_atom
-		if(!O.anchored && (O.move_resist < MOVE_FORCE_STRONG))
-			step(O, dir)
-		O.hitby(src, speed)
-
-	else if(isturf(hit_atom))
-		stop_throw()
-		var/turf/T = hit_atom
-		if(T.density)
-			if(bounce)
-				spawn(2)
-					step(src, turn(dir, 180))
-			if(isliving(src))
-				var/mob/living/M = src
-				M.turf_collision(T, speed)
+	else
+		var/old_throw_source = throw_source
+		hit_atom.hitby(src, speed)
+		if(bounce && hit_atom.density)
+			INVOKE_NEXT_TICK(src, PROC_REF(throw_bounce), hit_atom, old_throw_source)
 
 	SEND_SIGNAL(src, COMSIG_MOVABLE_IMPACT, hit_atom)
 
+///Bounces the AM off hit_atom
+/atom/movable/proc/throw_bounce(atom/hit_atom, turf/old_throw_source)
+	if(QDELETED(src))
+		return
+	if(!isturf(loc))
+		return
+	var/dir_to_proj = get_dir(hit_atom, old_throw_source)
+	if(ISDIAGONALDIR(dir_to_proj))
+		var/list/cardinals = list(turn(dir_to_proj, 45), turn(dir_to_proj, -45))
+		for(var/direction in cardinals)
+			var/turf/turf_to_check = get_step(hit_atom, direction)
+			if(turf_to_check.density)
+				cardinals -= direction
+		dir_to_proj = pick(cardinals)
```

Comment:
This is gonna runtime if both directions are somehow occupied. I don't know if that's actually possible with how diagonal movement works, mind you.

Reply:
yeah it shouldn't be possible, otherwise you wouldn't hit the thing in question to begin with.